devDocs/startupShutdownInstall.md

```diff
@@ -0,0 +1,82 @@
+1. For an installed copy, portable copy, installer:
+    1. An exiting instance of NVDA starting a new process (see shutting down procedures)
+    1. By running the exe (see cli options too).
+    This can be triggered by a user or external process such as an existing NVDA instance
+    - test: `startupShutdownNVDA.Starts`
+1. For source: eg runnvda.bat
+
+## NVDA can be shutdown by:
+
+1. UI within NVDA, with and without an ExitDialog prompt (uses `triggerNVDAExit`):
+    1. NVDA+q
+        - test: `startupShutdownNVDA.Quits from keyboard, Restarts`
+    1. An input gesture to restart
+    1. A prompt to restart from changing certain settings
+    1. Via the NVDA menu -> Exit
+        - test: `startupShutdownNVDA.Quits from menu`
+1. A process sending `WM_QUIT`, eg a new NVDA process starting (uses `WM_QUIT`)
+1. A handled crash (directly causes a new process to start, terminates unsafely)
+    - test: `startupShutdownNVDA.Restarts on crash, Restarts on braille crash`
+1. An unhandled crash (terminates unsafely)
+1. An external command which kills the process (terminates unsafely) 
+1. Windows shutting down (terminates unsafely) (uses `wx.EVT_END_SESSION`)
+
+## Exit hooks/triggers
+
+### In process: `triggerNVDAExit`
+* A function in the core module
+* Only executes the code once, uses a lock and flag to ensure this
+* Uses a queue on the main thread to queue a safe shutdown
+* Once the queued shutdown starts:
+    1. the updateCheck is terminated
+    1. watchdog is terminated
+    1. addons and the brailleViewer are terminated, so we can close all windows safely
+    1. All NVDA windows are closed
+    1. Now that windows are closed, a new NVDA instance is started if requested
+
+### Out of process: `WM_QUIT`
+* [a windows message](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-quit) which can be sent across processes which will force the main loop to exit and close most NVDA windows.
+* We subsequently run `triggerNVDAExit` to ensure that clean up code isn't missed, and pump the queue to execute it.
+* Changing to using a custom message, which would allow a custom handling (eg just `triggerNVDAExit`), requires compatibility with messaging older NVDA versions that are only aware of `WM_QUIT`.
+
+### Out of process: `wx.EVT_END_SESSION`
+- This is a [wxCloseEvent](https://docs.wxwidgets.org/3.0/classwx_close_event.html) triggered by a Windows session ending.
+- On `wx.EVT_END_SESSION`, we save the config and play the exit sound.
+- Other actions are not performed as we have limited time to perform an action for this event. With NVDA being killed off very last, Windows may include NVDA in the Block shutdown dialog, but the user won't be able to read it if we are shutting down.
+
+## Technical startup and exit notes
+
+## The Message Window
+
+The Windows API allows us to find a process by searching for a named window. To make the NVDA process searchable, NVDA creates a message window with a specific name.

```

Comment:
> named window

This is talking about 'window class' right? If so, perhaps "by searching for a named window (window class)"

Reply:
no, we use the standard window class name, we just give it a custom name attribute.

Reply:
Just a correction, we use the (old) standard wxWindow class name, `wxWindowClassNR`

Reply:
Microsoft generally just calls these objects "Windows" in the documentation, which makes things confusing.